modules/common/tls/tls.go

```diff
@@ -20,145 +20,403 @@ package tls
+	Ca `json:",inline"`
+}
+
+// APIService - API tls type which encapsulates for API services
+type APIService struct {
 	// +kubebuilder:validation:Optional
-	SecretName string `json:"secretName,omitempty"`
+	// Disabled TLS for the deployment of the service
+	Disabled *bool `json:"disabled,omitempty"`
+
+	// +kubebuilder:validation:optional
+	// +operator-sdk:csv:customresourcedefinitions:type=spec
+	// Public GenericService - holds the secret for the public endpoint
+	Public GenericService `json:"public,omitempty"`
+
+	// +kubebuilder:validation:optional
+	// +operator-sdk:csv:customresourcedefinitions:type=spec
+	// Internal GenericService - holds the secret for the internal endpoint
+	Internal GenericService `json:"internal,omitempty"`
+}
+
+// GenericService contains server-specific TLS secret or issuer
+type GenericService struct {
 	// +kubebuilder:validation:Optional
-	DisableNonTLSListeners bool `json:"disableNonTLSListeners,omitempty"`
+	// SecretName - holding the cert, key for the service
+	SecretName *string `json:"secretName,omitempty"`
 }
 
 // Ca contains CA-specific settings, which could be used both by services (to define their own CA certificates)
 // and by clients (to verify the server's certificate)
 type Ca struct {
-	// +kubebuilder:validation:Optional
-	CaSecretName string `json:"caSecretName,omitempty"`
+	// CaBundleSecretName - holding the CA certs in a pre-created bundle file
+	CaBundleSecretName string `json:"caBundleSecretName,omitempty"`
 }
 
-// TLS - a generic type, which encapsulates both the service and CA configurations
-type TLS struct {
-	Service *Service `json:"service"`
-	Ca      *Ca      `json:"ca"`
+// Service contains server-specific TLS secret
+// +kubebuilder:object:generate:=false
+type Service struct {
+	// SecretName - holding the cert, key for the service
+	SecretName string `json:"secretName"`
+
+	// CertMount - dst location to mount the service tls.crt cert. Can be used to override the default location which is /etc/tls/certs/<service id>.crt
+	CertMount *string `json:"certMount,omitempty"`
+
+	// KeyMount - dst location to mount the service tls.key  key. Can be used to override the default location which is /etc/tls/private/<service id>.key
+	KeyMount *string `json:"keyMount,omitempty"`
+
+	// CaMount - dst location to mount the CA cert ca.crt to. Can be used if the service CA cert should be mounted specifically, e.g. to be set in a service config for validation, instead of the env wide bundle.
+	CaMount *string `json:"caMount,omitempty"`
+}
+
+// Enabled - returns true if tls is configured for the public and internal
+func (a *APIService) Enabled(endpt service.Endpoint) bool {
+	switch endpt {
+	case service.EndpointPublic:
+		return (a.Disabled == nil || (a.Disabled != nil && !*a.Disabled)) && a.Public.SecretName != nil
+	case service.EndpointInternal:
+		return (a.Disabled == nil || (a.Disabled != nil && !*a.Disabled)) && a.Internal.SecretName != nil
+	}
+
+	return false
```

Comment:
I'm not convinced this is the best way to define the boolean that determines whether tls is enabled or not.

1. Its likely that - given our stance for not changing the tls status in adoption - we will have situations where tls is enabled only on the public interface.  I understand that this can be represented by disabled: false and setting the public secret name, but this also seems prone to error.  What if someone wanted tls (on either interface) but they forgot to provide the secret name.  In this scenario, the deployment would succeed but we would have no tls.  From a security point of view, I'd rather that the deployment failed, rather than accidentally exposing the interface.
2. A better implementation might be to have booleans for disable_public_tls and disable_internal_tls, which default to false and simply to check those booleans.  A separate check should also be done to confirm that the relevant secrets are set and if not, fail spectacularly.

Reply:
Or maybe it would fail in ValidateCertSecrets

Reply:
yes, `ValidateCertSecrets` should catch the situation where the secret is missing, or the content of the secret is wrong. I think we could just move the disable into the `GenericService` struct type, which is iiuc what you'd prefer, right? I'll give that a try how this impacts the current patches

Reply:
Agree.  Moving disable into GenericService would work.